### PR TITLE
Enable high availability through Knative leader election support

### PR DESCRIPTION
**Pull request by alan-ghelardi**
## What

- Enable high availability by making Watcher controllers leader aware. This allows the two forms of leader election (standard and statefulset-based) provided by Knative to be easily used in the Watcher.
- Add the threadiness flag to the Watcher binary. It allows users to tweak the number of workers allocated to process the controller's worker queue.
- Add the tekton-results-watcher Service. This is mostly a convenience to users who want to enable the statefulset-based leader election mechanism.
- Add ConfigMaps to customize logging, leader election and observability. The
config-leader-election ConfigMap allows users to tweak the leader election
mechanism. The other two ones make Results more homogeneous with Tekton
Pipeline project.

## Why

This pull request resolves https://github.com/tektoncd/results/issues/241, which
contains more details on why this feature is relevant. Essentially, it allows
users to tweak the Watcher controller to deal with higher loads more easily.

## Additional information

I tested those changes in a local Kind cluster and they're working as
expected. When the standard leader election mode is enabled, one replica
reconciles PipelineRuns and TaskRuns while the other one does nothing. When the
statefulset mode is enabled with three replicas for instance, all instances
reconcile different keys simultaneously.

### CODE REVIEW

File: pkg/watcher/reconciler/leaderelection/leader_election.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/reconciler"
+)
+
+// Lister is a generic signature of Lister.List functions, by allowing us to
+// support various listers in the NewLeaderAwareFuncs function below.
+type Lister[O metav1.Object] func(labels.Selector) ([]O, error)
```
Comment (dibyom) [inline pkg/watcher/reconciler/leaderelection/leader_election.go:28]: neat!

**Review by dibyom — APPROVED** [review]
cc @adambkaplan @vdemeester

**Review by vdemeester — APPROVED** [review]
/lgtm

**Review by khrm — COMMENTED** [review]
/kind feature
We need this label for release as discussed in the last WG call.

**Pull request merged**